### ISSUE TITLE
New release 2.0.2 does not validate against extra kwarg of WandbLogger when using CLI

### ISSUE THREAD

**Issue by w238liu** [Lightning-AI/lightning#17486]
### Bug description

After upgrading the recent release 2.0.2, the LightningCLI does not validate against extra kwargs, such as `group`, of the WandbLogger when using a config.yaml like `python main.py fit --config config.yaml`.

The minimal example is as follows:

`main.py`

```python
from lightning.pytorch.cli import LightningCLI
from lightning.pytorch.demos.boring_classes import DemoModel, BoringDataModule


def cli_main():
    cli = LightningCLI(                         # noqa
        DemoModel, BoringDataModule,
        save_config_kwargs={
            "config_filename": "pl_config.yaml",
            "overwrite": True
        }
    )


if __name__ == "__main__":
    cli_main()
```

`config.yaml`
```yaml
# lightning.pytorch==2.0.2
seed_everything: true
trainer:
  accelerator: auto
  strategy: auto
  devices: auto
  num_nodes: 1
  precision: 32-true
  logger:
    class_path: lightning.pytorch.loggers.WandbLogger
    init_args:
      save_dir: 'wandb'
      project: 'unittest'
      name: 'wandb_test'
      version: null
      log_model: all
      group: 'lightning 2.0.2'
      dir: 'wandb'
  callbacks: null
  fast_dev_run: false
  max_epochs: 100
  min_epochs: null
  max_steps: -1
  min_steps: null
  max_time: null
  limit_train_batches: null
  limit_val_batches: null
  limit_test_batches: null
  limit_predict_batches: null
  overfit_batches: 0.0
  val_check_interval: null
  check_val_every_n_epoch: 10
  num_sanity_val_steps: null
  log_every_n_steps: null
  enable_checkpointing: null
  enable_progress_bar: null
  enable_model_summary: null
  accumulate_grad_batches: 1
  gradient_clip_val: null
  gradient_clip_algorithm: null
  deterministic: null
  benchmark: null
  inference_mode: true
  use_distributed_sampler: true
  profiler: null
  detect_anomaly: false
  barebones: false
  plugins: null
  sync_batchnorm: false
  reload_dataloaders_every_n_epochs: 0
  default_root_dir: null
model:
  out_dim: 10
  learning_rate: 0.02
ckpt_path: null
```

Run with the command `python main.py fit --config config.yaml`, and it will error out with the following message:
```shell
usage: main.py [-h] [-c CONFIG] [--print_config[=flags]] {fit,validate,test,predict} ...
main.py: error: Parser key "trainer.logger":
  Does not validate against any of the Union subtypes
  Subtypes: (<class 'lightning.pytorch.loggers.logger.Logger'>, typing.Iterable[lightning.pytorch.loggers.logger.Logger], <class 'bool'>, <class 'NoneType'>)
  Errors:
    - Problem with given class_path 'lightning.pytorch.loggers.WandbLogger':
        'Configuration check failed :: No action for destination key "group" to check its value.'
    - Expected a <class 'collections.abc.Iterable'>
    - Expected a <class 'bool'>
    - Expected a <class 'NoneType'>
  Given value type: <class 'jsonargparse.namespace.Namespace'>
  Given value: Namespace(class_path='lightning.pytorch.loggers.WandbLogger', init_args=Namespace(name='wandb_test', save_dir='wandb', version=None, offline=False, dir='wandb', id=None, anonymous=None, project='unittest', log_model='all', experiment=None, prefix='', checkpoint_name=None, group='lightning 2.0.2'))
```

The same command works in the previous release `lightning==2.0.1.post0`.

### What version are you seeing the problem on?

2.0+

### How to reproduce the bug

See above.

### Error messages and logs

See above.

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA A10G
        - available:         True
        - version:           11.8
* Lightning:
        - lightning:         2.0.2
        - lightning-cloud:   0.5.34
        - lightning-utilities: 0.8.0
        - pytorch-lightning: 2.0.2
        - torch:             2.0.0
        - torchaudio:        2.0.0
        - torchmetrics:      0.11.4
        - torchqa:           0.2.1
        - torchvision:       0.15.0
* Packages:
        - aiohttp:           3.8.4
        - aiosignal:         1.3.1
        - alabaster:         0.7.13
        - anyio:             3.6.2
        - appdirs:           1.4.4
        - arrow:             1.2.3
        - async-timeout:     4.0.2
        - attrs:             23.1.0
        - babel:             2.12.1
        - beautifulsoup4:    4.12.2
        - blessed:           1.20.0
        - blockdiag:         3.0.0
        - boto3:             1.26.120
        - botocore:          1.29.120
        - brotlipy:          0.7.0
        - build:             0.10.0
        - certifi:           2022.12.7
        - cffi:              1.15.1
        - charset-normalizer: 2.0.4
        - click:             8.1.3
        - croniter:          1.3.14
        - cryptography:      39.0.1
        - dateutils:         0.6.12
        - deepdiff:          6.3.0
        - docker-pycreds:    0.4.0
        - docstring-parser:  0.15
        - docutils:          0.18.1
        - exceptiongroup:    1.1.1
        - fastapi:           0.88.0
        - filelock:          3.9.0
        - flake8:            6.0.0
        - frozenlist:        1.3.3
        - fsspec:            2023.4.0
        - funcparserlib:     1.0.1
        - gitdb:             4.0.10
        - gitpython:         3.1.31
        - gmpy2:             2.1.2
        - h11:               0.14.0
        - idna:              3.4
        - imagesize:         1.4.1
        - importlib-resources: 5.12.0
        - iniconfig:         2.0.0
        - inquirer:          3.1.3
        - itsdangerous:      2.1.2
        - jinja2:            3.1.2
        - jmespath:          1.0.1
        - jsonargparse:      4.21.0
        - lightning:         2.0.2
        - lightning-cloud:   0.5.34
        - lightning-utilities: 0.8.0
        - markdown-it-py:    2.2.0
        - markupsafe:        2.1.1
        - mccabe:            0.7.0
        - mdurl:             0.1.2
        - mkl-fft:           1.3.1
        - mkl-random:        1.2.2
        - mkl-service:       2.4.0
        - mpmath:            1.2.1
        - multidict:         6.0.4
        - mypy:              1.2.0
        - mypy-extensions:   1.0.0
        - networkx:          2.8.4
        - numpy:             1.23.5
        - ordered-set:       4.1.0
        - packaging:         23.1
        - pandas:            2.0.1
        - pathtools:         0.1.2
        - pillow:            9.4.0
        - pip:               23.0.1
        - pip-tools:         6.13.0
        - plotly:            5.14.1
        - pluggy:            1.0.0
        - protobuf:          4.22.3
        - psutil:            5.9.5
        - pycodestyle:       2.10.0
        - pycparser:         2.21
        - pydantic:          1.10.7
        - pyflakes:          3.0.1
        - pygments:          2.15.1
        - pyjwt:             2.6.0
        - pyopenssl:         23.0.0
        - pyproject-hooks:   1.0.0
        - pysocks:           1.7.1
        - pytest:            7.3.1
        - python-dateutil:   2.8.2
        - python-editor:     1.0.4
        - python-multipart:  0.0.6
        - pytorch-lightning: 2.0.2
        - pytz:              2023.3
        - pyyaml:            6.0
        - readchar:          4.0.5
        - readthedocs-sphinx-search: 0.3.1
        - requests:          2.28.1
        - rich:              13.3.4
        - s3fs:              0.4.2
        - s3transfer:        0.6.0
        - sentry-sdk:        1.21.0
        - setproctitle:      1.3.2
        - setuptools:        66.0.0
        - six:               1.16.0
        - smmap:             5.0.0
        - sniffio:           1.3.0
        - snowballstemmer:   2.2.0
        - soupsieve:         2.4.1
        - sphinx:            6.2.1
        - sphinx-autodoc-typehints: 1.23.0
        - sphinx-rtd-theme:  1.2.0
        - sphinxcontrib-applehelp: 1.0.4
        - sphinxcontrib-blockdiag: 3.0.0
        - sphinxcontrib-devhelp: 1.0.2
        - sphinxcontrib-htmlhelp: 2.0.1
        - sphinxcontrib-jquery: 4.1
        - sphinxcontrib-jsmath: 1.0.1
        - sphinxcontrib-qthelp: 1.0.3
        - sphinxcontrib-serializinghtml: 1.1.5
        - starlette:         0.22.0
        - starsessions:      1.3.0
        - sympy:             1.11.1
        - tenacity:          8.2.2
        - tomli:             2.0.1
        - torch:             2.0.0
        - torchaudio:        2.0.0
        - torchmetrics:      0.11.4
        - torchvision:       0.15.0
        - tqdm:              4.65.0
        - traitlets:         5.9.0
        - triton:            2.0.0
        - typeshed-client:   2.2.0
        - typing-extensions: 4.5.0
        - tzdata:            2023.3
        - urllib3:           1.26.15
        - uvicorn:           0.21.1
        - wandb:             0.15.0
        - wcwidth:           0.2.6
        - webcolors:         1.13
        - websocket-client:  1.5.1
        - websockets:        11.0.2
        - wheel:             0.38.4
        - yarl:              1.9.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.10.11
        - version:           #21~20.04.1-Ubuntu SMP Fri Aug 5 11:44:14 UTC 2022

</details>

cc @carmocca @mauvilsa

**Comment by carmocca** [comment]
I can reproduce the issue, but I also get the error with `2.0.1.post0`. It's possible that you accidentally upgraded your `jsonargparse` version and that triggered the error

Either way, the fix for this should be

```yaml
  logger:
    class_path: lightning.pytorch.loggers.WandbLogger
    init_args:
      save_dir: 'wandb'
      project: 'unittest'
      name: 'wandb_test'
      version: null
      log_model: all
      dir: 'wandb'
    dict_kwargs:
      group: 'lightning 2.0.2'
```

**Comment by mauvilsa** [comment]
I did notice this some days ago. The issue is not the version of jsonargparse. This was broken by https://github.com/Lightning-AI/lightning/pull/17385, and it wasn't noticed because it seems that the pull request checks skip this test https://github.com/Lightning-AI/lightning/blob/91cb4b9b87bdd59b0322dfad094e66a06ece4ff0/tests/tests_pytorch/test_cli.py#L1409

I have already implemented a fix in https://github.com/omni-us/jsonargparse/commit/b5cd693ab7f24b17c8540f8f6c83741e74adea78.

**Comment by carmocca** [comment]
Great! Please leave a comment with the version that fixes this when you release it. Thank you

**Comment by w238liu** [comment]
> I did notice this some days ago. The issue is not the version of jsonargparse. This was broken by #17385, and it wasn't noticed because it seems that the pull request checks skip this test
> 
> https://github.com/Lightning-AI/lightning/blob/91cb4b9b87bdd59b0322dfad094e66a06ece4ff0/tests/tests_pytorch/test_cli.py#L1409
> 
> I have already implemented a fix in [omni-us/jsonargparse@b5cd693](https://github.com/omni-us/jsonargparse/commit/b5cd693ab7f24b17c8540f8f6c83741e74adea78).

@mauvilsa Thanks for the fix!


> I can reproduce the issue, but I also get the error with `2.0.1.post0`. It's possible that you accidentally upgraded your `jsonargparse` version and that triggered the error
> 
> Either way, the fix for this should be
> 
> ```yaml
>   logger:
>     class_path: lightning.pytorch.loggers.WandbLogger
>     init_args:
>       save_dir: 'wandb'
>       project: 'unittest'
>       name: 'wandb_test'
>       version: null
>       log_model: all
>       dir: 'wandb'
>     dict_kwargs:
>       group: 'lightning 2.0.2'
> ```

@carmocca This way also works. I was wondering if it is recommended to pass extra kwargs within the `dict_kwargs` field rather than the `init_args` field in the future? I ask this because I would like to determine the format of the example config in my own library.

**Comment by mauvilsa** [comment]
@w238liu Just because there is a regression of a feature in some particular version of lightning, it is not a reason to just say provide all parameters in `dict_kwargs`.

The recommended approach to create a config is as you see in [prepare-a-config-file-for-the-cli](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_advanced.html#prepare-a-config-file-for-the-cli). For this example you would start with `main.py --trainer.logger=WandbLogger --print_config`. The output of this will include the defaults of all of the parameters that were successfully resolved in `init_args`. Recommending to provide them in `dict_kwargs` would be confusing. Are people supposed to move what they see in `init_args` to `dict_kwargs`? Also it becomes confusing that if you provide all parameters in `dict_kwargs`, in the [automatically save config](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_advanced.html#automatic-save-of-config) you will see them in `init_args`.

What you should explain is that if a parameter that is accepted does not appear in the `init_args` when doing `--print_config`, it means that the CLI was unable to resolve it. In these cases provide them in `dict_kwargs`. In the note at the end of [prepare-a-config-file-for-the-cli](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_advanced.html#prepare-a-config-file-for-the-cli) you can read a more detailed explanation.
